### DISCUSSION TITLE
thoughts-about-primary-constructors-3-pros-and-5-cons/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#420] [Announcements]
# thoughts-about-primary-constructors-3-pros-and-5-cons/

In this post I describe some of the ways I like to use primary constructors and some of the things I don't like about them

https://andrewlock.net/thoughts-about-primary-constructors-3-pros-and-5-cons/

**Comment by RichMercer** [comment]
The lack of `readonly` doesn't bother me too much with implicit capture, as reference types aren't strictly immutable anyway, but the naming convention bothers me. When I use a field in a method I want to use `_fieldName`, but I hate seeing `public ClassName(string _fieldName)`. I think camel case is the lesser of two evils, but my OCD is triggered every time!

**Comment by hahn-kev** [comment]
Is there a way to make the duplicate capture an error instead of a warning? we did this with nullable warnings and it's been pretty nice.

I agree that primary constructors pretty much break all naming conventions. In a record it should be PascalCase, for non captured parameters is should be camelCase, and for captured parameters it should be _ prefixed, at least to stay consistent with all previous dotnet conventions. But _fieldName just looks wrong in the constructor, always. And the PascalCase for records also feels wrong.

**Comment by RichardCrawshaw** [comment]
A number of things bug me about primary constructors: the lack of `readonly`, the naming convention discrepancies, the inability to use `this` as a prefix, and the restrictions around chaining constructors. Though in the "good" examples you mention I think they're excellent.

I find using a `this` prefix to all fields, properties and events far more beneficial than prepending '_' to their names. Not only do I find the underscore disrupts my reading of the variable name, when using `this` I can instruct the compiler to prompt for it to be included. Also '_' conflicts with code 

In terms of chaining constructors I would really like the primary constructor to be able to chain to a parameterless constructor that handles all the initialisation, such as initialising fields where the constructor of the type needs the current instance, or setting up event handlers, or even just a simple log trace message so I know my instance has been created. But this is stymied by all other constructors having to call the primary constructor, rather than the primary constructor needing to be in the chain.

**Comment by MarkJ1971** [comment]
I have a similar concern about the naming convention.  There was a discussion [in Sep 2023](https://github.com/dotnet/csharplang/discussions/7565) where some folks agreed with me & some didn't. IMHO the underscore convention is to identify the _scope_ of the identifier. It's very important to know the scope. It's less important to know whether it's a capture or a field.

I get that primary constructor parameters *are* parameters. I feel a better person for knowing that. But the captures can be used *in different methods* where they *aren't* parameters. So it is not the whole story to say they are "just" parameters. If they were "just" parameters, they would be useless! 

I feel the principle of least surprise is relevant. To me, names without underscores do look natural in the primary constructor, but unnatural in the rest of the class body. The primary constructor is short, simple and unlikely to cause issues. I spend far more time reading the rest of the class, which is longer, more complex, and more likely to cause issues. So it's more important for it to be understandable. 

I notice the C# style guide has this tip below. It seems misleading now. Typing _ doesn't show the PC captures. It's a symptom of this bigger problem.
>When editing C# code that follows these naming conventions in an IDE that supports statement completion, typing _ will show all of the object-scoped members.

Personally I would prefer one of these options 
 - add the underscore prefix to the primary constructor parameter 
 - drop the guidance that private fields should have underscore prefix. I believe Java doesn't have that convention and they seem to manage. IDEs could color the primary constructor captures differently: [Rider does](https://blog.jetbrains.com/dotnet/2023/11/23/primary-constructors-using-csharp-12-in-rider-and-resharper/#semantic-highlighting), don't know about others.
